src/core/utils/request/urlFromRequestOrCache.ts

```diff
@@ -0,0 +1,17 @@
+const urlCache = new Map<string, URL>()
+
+/**
+ * Returns the URL object for the given request. The URL object is cached so that
+ * subsequent calls to this function with the same request or requests to the same
+ * url will return the same URL object, to avoid multiple URL object allocations and
+ * parsing the same url multiple times.
+ */
+export function urlFromRequestOrCache(request: Request): URL {
+  const requestUrl = request.url
```

Comment:
We seem to be adopting this "look-in-cache-or-create" pattern quite extensively over these two pull requests. I think it may be worth exploring a tiny abstraction to make these actions more readable. 

```ts
function cached<V, K extends object>(producer: () => V, key: K, cache: WeakMap<K, V>): V {
  if (!cache.has(key)) {
    cache.set(key, producer())
  }
  
  return cache.get(key)!
}

// Usage.
const urlCache = new WeakMap<Request, URL>
export function cachedUrlFromRequest(request: Request) {
  return cached(() => new URL(request.url), request, urlCache)
}
```